### ISSUE TITLE
Rule Request: Bound method references

### ISSUE THREAD

**Issue by bradfol** [realm/SwiftLint#3693]
### New Issue Checklist

- [X] Updated SwiftLint to the latest version
- [X] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### New rule request

The idea for the rule would be to disallow usage of the language feature.

1. Why should this rule be added? Share links to existing discussion about what
   the community thinks about this.

    We have had a few instances of memory leaks in our app that were caused by bound method references, as described in this [implemented Swift Evolution proposal 0269](https://github.com/apple/swift-evolution/blob/main/proposals/0269-implicit-self-explicit-capture.md#bound-method-references). These are fairly difficult to debug, and it is not obviously clear to the engineer that writes the code that they are doing something that will likely cause a retain cycle.

2. Provide several examples of what _would_ and _wouldn't_ trigger violations.

```
class Test {
    var x = 0
    func execute(_ work: @escaping () -> Void) {
        work()
    }
    func method() {
        execute(inc) // This should cause a violation
        execute(self.inc) // This should also cause a violation
        execute({ self.inc() }) // This explicit use of a closure should not trigger a violation
    }
    func inc() {
        x += 1
    }
}
```
3. Should the rule be configurable, if so what parameters should be configurable?

   I don't think this needs configuration.
4. Should the rule be opt-in or enabled by default? Why?

   There is not yet general consensus on this rule.

**Comment by cmoulton** [comment]
Looks like it'd need to be an `AnalyzerRule` (which requires a separate compile & run from the regular linting pass https://github.com/realm/SwiftLint#analyze-experimental) https://realm.github.io/SwiftLint/Protocols/AnalyzerRule.html#/s:18SwiftLintFramework4RuleP8validate4fileSayAA14StyleViolationVGAA0aB4FileC_tF
